src/Features/CSharp/Portable/AddBraces/CSharpAddBracesDiagnosticAnalyzer.cs

```diff
@@ -86,5 +108,139 @@ public void AnalyzeNode(SyntaxNodeAnalysisContext context)
+                //   for (int i = 0; // <-- The initializer/condition/increment are on separate lines
+                //        i < 10;
+                //        i++)
+                //     SomeMethod();
+                return true;
+            }
+
+            if (statement.GetLastToken() != embeddedStatement.GetLastToken())
+            {
+                if (statement is IfStatementSyntax ifStatement && ifStatement.Statement == embeddedStatement)
+                {
+                    // The embedded statement is followed by an 'else' clause, which may span multiple lines without
+                    // triggering a braces requirement, such as this:
+                    //
+                    //   if (true)
+                    //     return;
+                    //   else          // <-- this else clause is two lines, but is not considered a multiline context
+                    //     return;
+                    //
+                    // ---
+                    // INTENTIONAL FALLTHROUGH
+                }
+                else
+                {
+                    var firstTokenAfterEmbeddedStatement = embeddedStatement.GetLastToken().GetNextToken();
+                    if (!FormattingRangeHelper.AreTwoTokensOnSameLine(firstTokenAfterEmbeddedStatement, statement.GetLastToken()))
+                    {
+                        // The part of the statement following the embedded statement does not fit on one line. Examples:
+                        //
+                        //   do
+                        //     SomeMethod();
+                        //   while (x < 0 ||    // <-- This condition is split across multiple lines.
+                        //          x > 10);
+                        return true;
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private static bool RequiresBracesToMatchContext(SyntaxNode statement, StatementSyntax embeddedStatement)
+        {
+            if (!statement.IsKind(SyntaxKind.IfStatement, SyntaxKind.ElseClause))
+            {
+                // 'if' statements are the only statements that can have multiple embedded statements which are
+                // considered relative to each other.
+                return false;
+            }
+
+            var topLevelContext = GetTopLevelContext(statement);
+            if (AnyPartUsesBraces(topLevelContext))
+            {
+                return true;
+            }
+
+            return false;
+        }
+
+        private static IfStatementSyntax GetTopLevelContext(SyntaxNode statement)
+        {
+            IfStatementSyntax result;
+            if (statement.IsKind(SyntaxKind.ElseClause))
+            {
+                result = (IfStatementSyntax)statement.Parent;
+            }
+            else
+            {
+                Debug.Assert(statement.IsKind(SyntaxKind.IfStatement));
+                result = (IfStatementSyntax)statement;
+            }
```

Comment:
consider adding the assert up front, then change this all to `var result = ... ? ... : ...`

Reply:
I prefer the current form. I don't like using the ternary operator for what I see as distinct operations.

Reply:
to me it was a single op "give me the relevant if-statement".  The need for this to be in a big-if-statement itself seemed mainly jsut to be able to assert something about the false-clause.  But, if you prefer this form, that's fine with me :)